src/pybindings/xcdf/xcdf.cpp

```diff
@@ -0,0 +1,227 @@
+#include <sstream>
+#include <string>
+#include <cstdio>
+#include <limits>
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+#include <pybind11/numpy.h>
+
+#include <xcdf/XCDF.h>
+#include "xcdf/XCDFFieldDescriptor.h"
+#include "xcdf/XCDFFile.h"
+
+namespace py = pybind11;
+
+std::string getVersion()
+{
+    std::stringstream ss;
+    ss << XCDF_MAJOR_VERSION
+       << "." << XCDF_MINOR_VERSION
+       << "." << XCDF_PATCH_VERSION;
+    return ss.str();
+}
+
+int write_test_file()
+{
+    XCDFFile f("test_file_for_pybindings.xcd", "w");
+
+    XCDFUnsignedIntegerField field1 =
+        f.AllocateUnsignedIntegerField("field1", 1);
+    XCDFUnsignedIntegerField field2 =
+        f.AllocateUnsignedIntegerField("field2", 1, "field1");
+    XCDFFloatingPointField field3 =
+        f.AllocateFloatingPointField("field3", 0.1);
+    XCDFFloatingPointField field4 =
+        f.AllocateFloatingPointField("field4", 0.1);
+    XCDFFloatingPointField field5 =
+        f.AllocateFloatingPointField("field5", 0.1);
+    XCDFUnsignedIntegerField field6 =
+        f.AllocateUnsignedIntegerField("field6", 1);
+    XCDFFloatingPointField field7 =
+        f.AllocateFloatingPointField("field7", 0.);
+    // 2D vector
+    XCDFUnsignedIntegerField field8 =
+        f.AllocateUnsignedIntegerField("field8", 1, "field2");
+    // 3D vector
+    XCDFFloatingPointField field9 =
+        f.AllocateFloatingPointField("field9", 0.5, "field8");
+    // Header alias
+    f.CreateAlias("testAlias", "field1 + 1");
+
+    field1 << 2;
+    field2 << 1 << 1;
+    field3 << 0.1;
+    field4 << 5.;
+    field5 << 5.;
+    field6 << 0xDEADBEEFDEADBEEFULL;
+    field7 << 0.12;
+    field8 << 2 << 1;
+    field9 << 1. << 2. << 3.;
+
+    std::cout << f.Write() << std::endl;
+
+    field1 << 2;
+    field2 << 1 << 3;
+    field3 << 0.3;
+
+    // Write a NaN into field4
+    field4 << std::numeric_limits<double>::signaling_NaN();
+
+    // Write an inf into field5
+    field5 << std::numeric_limits<double>::infinity();
+
+    field6 << 0xDEADBEEFDEADBEEFULL;
+
+    field7 << 0.12;
+
+    field8 << 2 << 2 << 1 << 1;
+    field9 << 1. << 2. << 3. << 4. << 5. << 6.;
+
+    std::cout << f.Write() << std::endl;
+
+    for (int k = 0; k < 1000; k++)
+    {
+        field1 << 2;
+        field2 << 1 << 3;
+        field3 << 0.3;
+        field4 << std::numeric_limits<double>::signaling_NaN();
+        field5 << std::numeric_limits<double>::infinity();
+        field6 << 0xDEADBEEFDEADBEEFULL;
+        field7 << 0.12;
+        field8 << 2 << 2 << 1 << 1;
+        field9 << 1. << 2. << 3. << 4. << 5. << 6.;
+        f.Write();
+    }
+
+    f.AddComment("test file");
+
+    // Trailer alias
+    f.CreateAlias("testTrailerAlias", "double(testAlias + 2)");
+
+    f.Close();
```

Comment:
RAII. This should not be necessary but be call `Close` here, it should be called on exit by the destructor.

Reply:
My C++ is quite rusty, but I don't see an automatic call to a destructor in the code of XCDF.cc...

Here when I call

`XCDFFile f("test_file_for_pybindings.xcd", "w");`

it should just mean `XCDFFile::Init`

`write_test_file` is basically the first half of `tests/SimpleTest.cc` and also there the Close method is explicitly called